File: Session13/Java/KClosestPointsToOrigin.java
https://leetcode.com/problems/k-closest-points-to-origin/

class Solution {
    public int[][] kClosest(int[][] points, int K) {
        
        PriorityQueue<int[]> heap = new PriorityQueue<>
            ((a, b) -> (b[0]*b[0] + b[1]*b[1] - (a[0]*a[0] + a[1]*a[1])));
        
        for (int [] point: points) {
            heap.add(point);
            if (heap.size() > K) {
                heap.poll();   
            }
        }
        
        int [][] result = new int[K][2];
        while (K-- > 0) {
            result[K] = heap.poll();
        } 
        return result;
    }
}

File: Session13/C++/KClosestPointsToOrigin.cpp
https://leetcode.com/problems/k-closest-points-to-origin/

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        
        int n = points.size();
        priority_queue<pair<int, pair<int, int>>> pq; // max heap
        
        for(int i=0; i<n; i++) {
            int dist = points[i][0]*points[i][0] + points[i][1]*points[i][1];
            pq.push({dist, {points[i][0], points[i][1]}});
            if(pq.size()>k) {
                pq.pop();
            }
        }
        
        vector<vector<int>> ans;
        while(!pq.empty()) {
            pair<int, int> coor = pq.top().second;
            ans.push_back({coor.first, coor.second});
            pq.pop();
        }
        
        return ans;   
    }
};